### PR TITLE
[SPARK-16380][EXAMPLES] Update SQL examples and programming guide for Python language binding

### PR DESCRIPTION
**Pull request by liancheng**
This PR is based on PR #14098 authored by @wangmiao1981.
## What changes were proposed in this pull request?

This PR replaces the original Python Spark SQL example file with the following three files:
- `sql/basic.py`
  
  Demonstrates basic Spark SQL features.
- `sql/datasource.py`
  
  Demonstrates various Spark SQL data sources.
- `sql/hive.py`
  
  Demonstrates Spark SQL Hive interaction.

This PR also removes hard-coded Python example snippets in the SQL programming guide by extracting snippets from the above files using the `include_example` Liquid template tag.
## How was this patch tested?

Manually tested.

### CODE REVIEW

**Comment by liancheng** [comment]
@JoshRosen Would you mind to have a look at this? Thanks!

**Comment by rxin** [comment]
Merging in master/2.0.

File: docs/sql-programming-guide.md [hunk 1]
```diff
@@ -79,7 +79,7 @@ The entry point into all functionality in Spark is the [`SparkSession`](api/java
 
 The entry point into all functionality in Spark is the [`SparkSession`](api/python/pyspark.sql.html#pyspark.sql.SparkSession) class. To create a basic `SparkSession`, just use `SparkSession.builder`:
 
-{% include_example init_session python/sql.py %}
+{% include_example init_session python/sql/basic.py %}
```
Comment (wangmiao1981) [inline docs/sql-programming-guide.md]: The file name is not consistent with Scala and Java version. The file names are SparkSQLExample.scala and SparkSQLExample.java. The Hive and Data Source examples file names are not consistent either.

File: examples/src/main/python/sql/basic.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    # +-------+
+    # |   name|
+    # +-------+
+    # |Michael|
+    # |   Andy|
+    # | Justin|
+    # +-------+
+
+    # Select everybody, but increment the age by 1
+    df.select(df['name'], df['age'] + 1).show()
```
Comment (wangmiao1981) [inline examples/src/main/python/sql/basic.py]: Do you want to use `col('...')`. I have tested it and it works.

File: examples/src/main/python/sql/datasource.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    # |        address|name|
+    # +---------------+----+
+    # |[Columbus,Ohio]| Yin|
+    # +---------------+----+
+    # $example off:json_dataset$
+
+if __name__ == "__main__":
+    spark = SparkSession \
+        .builder \
+        .appName("PythonSQL") \
```
Comment (wangmiao1981) [inline examples/src/main/python/sql/datasource.py]: The basic example also has the name "PythonSQL". Do you want to give it a unique name?

File: docs/sql-programming-guide.md (hunk 1 above)
Comment (liancheng) [inline docs/sql-programming-guide.md]: For Scala and Java, it's a convention that the file name should be the same as the (major) class defined in the file, while camel case file name doesn't conform to Python code convention. You may check other PySpark file names in the repo as a reference.

File: examples/src/main/python/sql/basic.py (hunk 2 above)
Comment (liancheng) [inline examples/src/main/python/sql/basic.py]: Yea, I know I brought up this issue, but it is still in question... Although `df['...']` has potential issue with self-join, it is the way Pandas DataFrame works. Considering we've tried to workaround various self-join issues within Catalyst, now I tend to preserve it as is. Maybe we'll deprecate this syntax later.

File: examples/src/main/python/sql/datasource.py (hunk 3 above)
Comment (liancheng) [inline examples/src/main/python/sql/datasource.py]: Nice catch! Will fix it in a follow-up PR, thanks!

**Pull request closed without merging**